pkg/policy/rule.go

```diff
@@ -66,6 +71,30 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
+
+	if filter.addEndpoints(endpoints) && r.NumRules() == 0 {
+		// skip this policy as it is already covered and it does not contain L7 rules
+		return 1, nil
+	}
+
+	// Rule we are merging with existing L4Filter has no L7 rules, but applies
+	// to the same L4 port-protocol tuple, which *does* have L7 rules. If a rule
+	// applying to the same port/protocol has no L7 rules, then that equates to
+	// allowing all on L7, so just remove existing L7-related metadata for this
+	// port-proto tuple. Or, if we already have a filter which allows all L4, yet
+	// are trying to add a rule which restricts on L7, do not restrict on L7.
+	if r.NumRules() == 0 || filter.L7Parser == "" {
+		for k := range filter.L7RulesPerEp {
+			delete(filter.L7RulesPerEp, k)
+		}
+		filter.L7Parser = ""
+		filter.DerivedFromRules = append(filter.DerivedFromRules, ruleLabels)
```

Comment:
Here, you're preserving the old rules. I don't think you should, because that won't always be correct.
E.g. if rule A is `allow ingress from app=abc to port 80 HTTP GET /foo`, and rule B is `allow ingress from app=abc to port 80`, there should be no trace of rule A, because it would be incorrect to report that rule A matches a flow with a different URL such as `from app=abc to port 80 HTTP GET /bar`. In this case, I think you should only keep a reference to rule B.

Reply:
I think the 'DerivedFromRules' sort of gets broken here with the condition you apply. There are multiple cases to consider here:

* Both the rule which we are merging into this L4Filter has no rules *and* the existing L7Parser is empty (no parser; in that case, append the ruleLabels from the rule we are merging into this L4Filter.
* The L4Filter has multiple DerivedFromRules and has a non-empty L7Parser currently (meaning that some rules which were merged into this L4Filter apply at L7), but now we are importing a rule with no L7Rules. What ruleLabels in this case do we remove from the DerivedFromRules list? We can't be sure that these rules are at this location in the code path unless we get the rule itself from the policy repository and analyze it here - at least I think this is the case. What do you think?
* The case you specified, where if we have previously imported rule B, and then import rule A, then we wouldn't add rule A to the DerivedRules because it is shadowed by B.

I view `DerivedFromRules` more of a way of seeing what rules went into the computation of this L4Filter. They could be very well shadowed by other rules; I think it's up to the user to make that distinction. I actually think updating `DerivedFromRules` with rules that don't end up doing anything reveals to the user that they have a rule which is functionally useless.

Reply:
As a user I would expect that all rules that contributed to allowing something should be listed. We should refer to `policy trace` for details what gets shadowed where. I would want to know what is affected if I remove a rule. I would look at DerivedRules for that.

Reply:
OK, makes sense. I was confusing this with `policy trace`.